debian/man/zimsplit.1
.TH ZIMSPLIT "1" "July 2020" "zim-tools" "User Commands"
.SH NAME
zimsplit \- split a ZIM file into smaller parts
.SH SYNOPSIS
\fBzimsplit\fR [options] zimfile\fR
.SH DESCRIPTION
.TP
zimfile
ZIM file to split
.TP
\fB\-o\fR PREFIX
Prefix of output file parts
.TP
\fB\-\-size=\fISIZE\fR
Maximum size of each part. SIZE must be a positive integer, optionally followed
by a decimal unit (KB, MB, GB, TB, PB, or EB) or a binary unit (KiB, MiB, GiB,
TiB, PiB, or EiB). Units are case-insensitive. A value without a unit is
interpreted as bytes. SIZE must be smaller than the input file. The default is
2GiB.
.TP
\fB\-\-force\fR
create ZIM parts even if it is impossible to have all part size smaller than requested
.TP
\fB\-v\fR, \fB\-\-version\fR
print software version



ChangeLog
zim-tools 3.8.0
===============

 * zimsplit: Fixed (large) --size option (@veloman-yunkan #520)
 * zimcheck: Validation of MIME type for reserved metada (@veloman-yunkan #530)
 * zimcheck: Improve memory mgmt around message logging (@Pranavjeet-Naidu #500)
 * zimcheck: Introduce --quick option to avoid reporting all link error for each page (@Pranavjeet-Naidu #505)
 * zimcheck: Overhaul of link normalization to fix many links related bug (@veloman-yunkan #510 #511)
 * zimcheck: Count graphemes instead of unicode code point in metadata length checks (@Pranavjeet-Naidu #499)
 * zimcheck: Properly parse all <script> src attributes (@veloman-yunkan #525)
 * zimcheck: Classify arbitrary URI schemes as external (@tomatotomata #518)
 * zimwriterfs: Improve the refresh URL parsing (@kelson42 #515)
 * Update CI/CD workflows (@veloman-yunkan #532, @kelson42 #503)
 * Compilation relies now solely on libmagic.pc (@kelson42 #504)

zim-tools 3.7.0
===============

 * zimwriterfs: Better handling of file extension (case) (@veloman-yunkan #494)
 * zimcheck: Properly check the `M/Counter` value (@Pranavjeet-Naidu #497)
 * zimcheck: switch to new API for illustrations (@veloman-yunkan #476)
 * Various updates in the CI (@kelson42 #472 #483 #493)
 * Requires now libzim 9.4, instead of 9.2 (@kelson42 #480)

zim-tools 3.6.0
===============

 * zimrecreate: Do not copy entry in 'W' namespace (@mgautierfr #452)
 * zimwriterfs: Move function declarations into correct header (@ThisIsFineTM #451)
 * zimwriterfs: Fix usage sample command (@benoit74 #435)
 * zimwriterfs/zimcheck: Fix buffer overrun in link normalisation (@ThisIsFineTM #446 #450)
 * Few fixes for FreeBSD compilation (@OlCe2 #445)
 * Stricter libzim dependency version check (@kelson42 #431 #437)
 * Fix gcc 14.2 LTO compilation bug in metadata-test (@ThisIsFineTM #460)
 * Update CI/CD workflows (@kelson42 #436)

zim-tools 3.5.0
===============

 * Add support of new Debian based distribution (@kelson42 #424)
 * Add MW Windows base CI/CD (@mgauthier #425)

zim-tools 3.4.2
===============

 * zimwriterfs: Support short option `-a` (tags) (@kelson42 #409)
 * zimwriterfs: "Name", "Title" and "LongDescription" are mandatory metadata (@kelson42 #410 #411)
 * zimcheck: Do not run check on content if zim integrity test is failing (@mgautierfr #413)
 * zimrecreate: Do not try to add twice metadata (and fix return code) (@mgautiefr #419)

zim-tools 3.4.1
===============

  * zimsplit: stop creating trailing empty chunk (@mgautierfr #402)
  * zimdump: Respecting the --ns option in zimdump show (@veloman-yunkan #316)
  * Introduce max libzim version check in compilation (@kelson42 #397)

zim-tools 3.4.0
===============

  * zimwriterfs: Fix regression of Redirect entries not findable in title suggestions (@mgautierfr #388)
  * zimcheck: Do not use std::optional (@mgautierfr #392)
  * zimcheck: Do not detect ZIM clone entry as duplicated content (@mgautierfr #379)

zim-tools 3.3.0
===============

  * zimcheck: Properly deal with HTML entities in URL (@veloman-yukan #378)
  * zimcheck: Stop complaining about empty href (@BartlomiejZalas #367)
  * zimcheck: Fix handling of bad or missing magic file (@rgaudin #374)
  * zimwriterfs, zimcheck: Fix illustration check (@mgautierfr #369)
  * zimwriterfs: Stop compressing illustrations (@mgautierfr #355)
  * zimwriterfs: Properly deal with wrong amount of columns in redirect file (@mgautierfr #360)
  * Fix Debian package compilation on arm64 (@mgautierfr #369)
  * Full code base compliancy with C++17 (@mgautierfr #368)
  * Container images for ARM CPU architecture (@rgaudin #351)
  * Multiple CI/CD improvements (@kelson42)

zim-tools 3.2.0
===============

  * Introduce a "Metadata table" to centralize information about metadata and check them (@veloman-yunkan #339)
  * Add check for description and longDescription (@Onyx2406 #333)
  * Do not use `--static` option when compiling on MacOs (@mgautierfr #342)
  * Add support for SIP uri scheme (@kelson42 #329)
  * Various CI improvement (@kelson42)

zim-tools 3.1.3
===============

  * zimbench: Fix command line parsing (@veloman-yunkan #323)

zim-tools 3.1.2
===============

  * zimdump: Create correct redirection when dumping zim with old namespace
    scheme (@mgautierfr #306)
  * zimcheck: Correctly handle protocol-relative absolute url as absolute url (@veloman-yunkan #307)
  * zimcheck: Fix `J` flag at arguments parsing (@JameelKaisar #320)
  * zimcheck: Limit the size of the task queue, which reduce the memory needed (@veloman-yunkan #309)
  * zimcheck: Better and faster redirect loop check (@veloman-yunkan #312)
  * zimdump: Improve indentation output of `zimdump list` (@kelson42 #301)
  * zimwriterfs: Improve an error message (@kelson42 #302)
  * CI: Fix CI compilation (@mgautierfr #313)


zim-tools 3.1.1
===============

 * zimdump: Correctly print the main page (@mgautierfr #293)
 * zimcheck: Correctly print spend time to avoid performance test to fail on slow
   hardware (@veloman-yunkan #288)
 * zimwriterfs: Fix usage message indentation (@kelson42 #291)
 * Cross compilation: Correctly detect host machine (@kelson42 #290)
 * CI: Fix libzim-dev deps (@legoktm #284)
 * CI: Remove Ubuntu Hirsute (@legoktm #286)

zim-tools 3.1.0
===============

 * zimcheck: Fix json output (Use `"` instead of `'`) for strings
   (@mgautierfr #281)
 * zimcheck: Avoid undefined behavior (and crash) when testing link's content.
   (@mgautierfr #280)
 * The `--version` option now also print the version of all dependencies (@kelson42, #275)
 * zimcheck: Add the zimcheck version in the output log (@kelson42, #274)
 * CI: Readd build on Ubuntu Impish (@kelson42, #267)
 * CI: Add build on Ubuntu Jammy (@legoktm, #273)


zim-tools 3.0.0
===============

 * zimwriterfs: Do not allow zimwriterfs to overwrite a existing zim file
   (@asashnov #193)
 * zimcheck: Make zimcheck print json output with option `--json`
   (@veloman-yunkan #206)

src/zimsplit.cpp
/*
 * Copyright (C) 2017 Matthieu Gautier.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU  General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

#include <fstream>
#include <sstream>
#include <stdexcept>

#define ZIM_PRIVATE
#include <zim/archive.h>

#include <docopt/docopt.h>

#include "version.h"
#include "zimsplit_size.h"

#define BUFFER_SIZE 4096

const zim::size_type DEFAULT_PART_SIZE = 2147483648;

class ZimSplitter
{
  private:
    zim::Archive archive;
    const std::string prefix;
    zim::size_type maxPartSize;

    char first_index, second_index;

    std::ifstream ifile;
    std::ofstream ofile;
    std::string part_name;
    zim::size_type currentPartSize;
    char* batch_buffer;

  public:
    ZimSplitter(const std::string& fname, const std::string& out_prefix, zim::size_type maxPartSize)
      : archive(fname),
        prefix(out_prefix),
        maxPartSize(maxPartSize),
        first_index(0),
        second_index(0),
        ifile(fname, std::ios::binary),
        currentPartSize(0)
      {
        validatePartSize(maxPartSize, archive.getFilesize());
        batch_buffer = new char[BUFFER_SIZE];
    }

    ~ZimSplitter() {
        close_file();
        ifile.close();
        delete batch_buffer;
    }

    std::string get_new_suffix() {
        auto out = std::string(1, 'a'+first_index) + std::string(1, 'a'+second_index++);
        if (second_index > 25) {
            first_index++;
            second_index = 0;
        }
        if (first_index > 25) {
            std::cerr << "To many parts" << std::endl;
            exit(-1);
        }
        return out;
    }

    void close_file() {
        if (currentPartSize > maxPartSize) {
           std::cout << "WARNING: Part " << part_name << " is bigger than max part size."
            << " (" << currentPartSize << ">" << maxPartSize << ")" << std::endl;
        }
        ofile.close();
    }

    void new_file() {
        close_file();
        part_name = prefix + get_new_suffix();
        std::cout << "opening new file " << part_name << std::endl;
        ofile.open(part_name, std::ios::binary);
        currentPartSize = 0;
    }

    void copy_out(zim::size_type size) {
        while (size > 0) {
           auto size_to_copy = std::min<zim::size_type>(size, BUFFER_SIZE);
           ifile.read(batch_buffer, size_to_copy);
           if (!ifile) {
               throw std::runtime_error("Error while reading zim file");
           }
           ofile.write(batch_buffer, size_to_copy);
           if (!ofile) {
               throw std::runtime_error("Error while writing zim part");
           }
           currentPartSize += size_to_copy;
           size -= size_to_copy;
        }
    }

    std::vector<zim::offset_type> getOffsets() {
        std::vector<zim::offset_type> offsets;
        auto nbClusters = archive.getClusterCount();
        for(zim::offset_type i=0;i<nbClusters;i++) {
            offsets.push_back(archive.getClusterOffset(i));
        }
        offsets.push_back(archive.getFilesize());
        std::sort(offsets.begin(), offsets.end());
        return offsets;
    }

    void run() {
        new_file();

        auto offsets = getOffsets();

        zim::offset_type last(0);
        for(auto offset:offsets) {
            auto chunkSize = offset-last;
            if (currentPartSize > 0 && currentPartSize + chunkSize > maxPartSize) {
                    new_file();
            }
            copy_out(chunkSize);
            last = offset;
        }
    }

    bool check() {
        bool error = false;

        auto offsets = getOffsets();

        zim::offset_type last(0);
        for(auto offset:offsets) {
            auto chunkSize = offset-last;
            if (chunkSize > maxPartSize) {
                // One part is bigger than what we want :/
                // Still have to write it.
                std::cout << "The part (probably a cluster) is to big to fit in one part." << std::endl;
                std::cout << "    size is " << chunkSize << "(" << offset << "-" << last << ")." << std::endl;
                error = true;
            }
            last = offset;
        }
       return error;
    }
};

static const char USAGE[] = R"(
    zimsplit splits smartly a ZIM file in smaller parts.

Usage:
    zimsplit [--prefix=PREFIX] [--force] [--size=SIZE] <file>
    zimsplit --version

Options:
    --prefix=PREFIX     Prefix of output file parts. Default: <file>
    --size=SIZE         Maximum part size in bytes, or with a decimal unit (KB, MB, ...)
                        or binary unit (KiB, MiB, ...). Units are case-insensitive.
                        SIZE must be smaller than the input file. Default: 2GiB
    --force             Create zim parts even if it is impossible to have all part size smaller than requested
    -h, --help          Show this help message
    --version           Show zimsplit version.
)";

int main(int argc, char* argv[])
{
  try
  {
    std::ostringstream versions;
    printVersions(versions);
    auto args = docopt::docopt(USAGE,
                              {argv + 1, argv + argc},
                              true,
                              versions.str());

    std::string prefix = args["<file>"].asString();
    if (args["--prefix"])
        prefix = args["--prefix"].asString();

    zim::size_type size = DEFAULT_PART_SIZE;
    if (args["--size"])
        size = parseByteSize(args["--size"].asString());

    // initalize app
    ZimSplitter app(args["<file>"].asString(), prefix, size);

    if (!args["--force"] && app.check()) {
        std::cout << "Creation of zim parts canceled because of previous errors." << std::endl;
        std::cout << "Use --force option to create zim parts anyway." << std::endl;
        return -1;
    }

    app.run();
  }
  catch (const std::exception& e)
  {
    std::cerr << e.what() << std::endl;
    return -2;
  }
  return 0;
}

